File: markdown_gate/scanner.py
from __future__ import annotations

from dataclasses import dataclass
import re
from typing import Iterable

from .config import GateConfig
from .model import Document, Finding, Severity


@dataclass(frozen=True)
class PatternRule:
    rule: str
    regex: re.Pattern[str]
    message: str


PATTERNS = [
    PatternRule(
        "revision_trace",
        re.compile(
            r"(已移除|已删除|已调整|已修正|修正为|调整为|改成|"
            r"has been removed|was removed|removed from|updated from|fixed by|"
            r"dropped\b)",
            re.I,
        ),
        "Looks like revision history rather than final-state documentation.",
    ),
    PatternRule(
        "rejected_prior_solution",
        re.compile(
            r"(之前方案|上一版|旧方案|原方案|历史方案|这里不采用|不采用该方案|"
            r"previous (design|approach|version)|old (design|approach)|"
            r"rejected (design|approach|alternative)|not adopted|"
            r"formerly\b|former version|prior version|legacy (field|design|approach)|"
            r"deprecated|superseded|replaced by|renamed from|originally)",
            re.I,
        ),
        "Looks like rejected or prior design context.",
    ),
    PatternRule(
        "self_correction",
        re.compile(
            r"(更正|纠正|修正说明|前面.*错误|此前.*错误|不是.*而是|"
            r"correction|corrected)",
            re.I,
        ),
        "Looks like self-correction language.",
    ),
    PatternRule(
        "internal_authoring_process",
        re.compile(
            r"(为了避免.*误解|为了避免.*上一|为避免.*再次|按你的反馈|根据反馈|"
            r"as requested|based on feedback|to avoid confusion|draft residue|"
            r"drafting history|draft note|草稿痕迹|草稿说明)",
            re.I,
        ),
        "Looks like internal authoring process leaking into the document.",
    ),
    PatternRule(
        "negative_constraint",
        re.compile(
            r"(\bdo not\b|\bdon't\b|\bmust not\b|\bno\b(?!\s+(?:longer|more)\s+than)|"
            r"不要|不得|禁止)",
            re.I,
        ),
        "Negative constraint detected; verify it belongs in this document type and section.",
    ),
]


HEADING_RE = re.compile(r"^(#{1,6})\s+(.+?)\s*$")
FENCE_RE = re.compile(r"^\s*(```|~~~)")
ALLOWED_SECTION_RULES = {
    "revision_trace",
    "rejected_prior_solution",
    "negative_constraint",
}


@dataclass(frozen=True)
class Segment:
    line: int
    column: int
    start: int
    text: str


@dataclass(frozen=True)
class ProseBlock:
    text: str
    segments: tuple[Segment, ...]
    sections: tuple[str, ...]

    @property
    def section(self) -> str | None:
        return self.sections[-1] if self.sections else None


def scan_documents(documents: Iterable[Document], config: GateConfig) -> list[Finding]:
    findings: list[Finding] = []
    for document in documents:
        findings.extend(scan_document(document, config))
    return findings


def scan_document(document: Document, config: GateConfig) -> list[Finding]:
    findings: list[Finding] = []
    for block in iter_prose_blocks(document.text):
        for pattern in PATTERNS:
            match = pattern.regex.search(block.text)
            if not match:
                continue

            if config.is_allowed_section_path(document.doc_type, list(block.sections)):
                if pattern.rule in ALLOWED_SECTION_RULES:
                    continue
                # Allowed sections still surface very process-heavy wording, but
                # ordinary alternatives/constraints do not block final copy.
                severity = config.severity_for(document.doc_type, pattern.rule)
                if severity <= Severity.SUGGESTION:
                    continue
            else:
                severity = config.severity_for(document.doc_type, pattern.rule)

            if severity == Severity.OFF:
                continue

            line_number, column = _locate_match(block, match.start())
            findings.append(
                Finding(
                    path=document.path,
                    doc_type=document.doc_type,
                    line=line_number,
                    column=column,
                    rule=pattern.rule,
                    severity=severity,
                    message=pattern.message,
                    excerpt=block.text.strip(),
                    section=block.section,
                    sections=block.sections,
                )
            )

    return findings


def iter_prose_blocks(text: str) -> Iterable[ProseBlock]:
    section_stack: list[tuple[int, str]] = []
    segments: list[Segment] = []
    in_fence = False

    def flush() -> ProseBlock | None:
        nonlocal segments
        if not segments:
            return None
        block_text = " ".join(segment.text for segment in segments)
        offset = 0
        normalized_segments: list[Segment] = []
        for segment in segments:
            normalized_segments.append(
                Segment(
                    line=segment.line,
                    column=segment.column,
                    start=offset,
                    text=segment.text,
                )
            )
            offset += len(segment.text) + 1
        segments = []
        return ProseBlock(
            text=block_text,
            segments=tuple(normalized_segments),
            sections=tuple(title for _, title in section_stack),
        )

    for line_number, line in enumerate(text.splitlines(), start=1):
        stripped = line.strip()

        if FENCE_RE.match(line):
            block = flush()
            if block:
                yield block
            in_fence = not in_fence
            continue

        if in_fence:
            continue

        heading = HEADING_RE.match(line)
        if heading:
            block = flush()
            if block:
                yield block
            level = len(heading.group(1))
            title = _strip_markdown_title(heading.group(2))
            while section_stack and section_stack[-1][0] >= level:
                section_stack.pop()
            section_stack.append((level, title))
            continue

        if not stripped or _looks_like_table_row(stripped):
            block = flush()
            if block:
                yield block
            continue

        column = len(line) - len(line.lstrip()) + 1
        segments.append(Segment(line=line_number, column=column, start=0, text=stripped))

    block = flush()
    if block:
        yield block


def _locate_match(block: ProseBlock, offset: int) -> tuple[int, int]:
    for segment in block.segments:
        end = segment.start + len(segment.text)
        if segment.start <= offset <= end:
            return segment.line, segment.column + max(0, offset - segment.start)
    if block.segments:
        last = block.segments[-1]
        return last.line, last.column
    return 1, 1


def _looks_like_table_row(stripped: str) -> bool:
    if stripped.startswith("|") and stripped.endswith("|"):
        return True
    return bool(re.fullmatch(r"\|?\s*:?-{3,}:?\s*(\|\s*:?-{3,}:?\s*)+\|?", stripped))


def _strip_markdown_title(title: str) -> str:
    return title.strip().strip("#").strip()

File: src/scanner.ts
import { type GateConfig, isAllowedSectionPath, severityFor } from "./config.js";
import type { Document, Finding, Severity } from "./model.js";

const HEADING_RE = /^(#{1,6})\s+(.+?)\s*$/;
const FENCE_RE = /^\s*(```|~~~)/;

interface Segment {
  line: number;
  column: number;
  start: number;
  text: string;
}

interface ProseBlock {
  text: string;
  segments: Segment[];
  sections: string[];
  section?: string;
}

export function scanDocuments(documents: Document[], config: GateConfig): Finding[] {
  return documents.flatMap((document) => scanDocument(document, config));
}

export function scanDocument(document: Document, config: GateConfig): Finding[] {
  const findings: Finding[] = [];
  for (const block of iterProseBlocks(document.text)) {
    for (const rule of config.rules) {
      const match = rule.regex.exec(block.text);
      if (!match || match.index === undefined) {
        continue;
      }

      const severity: Severity = severityFor(config, document.docType, rule.id);
      if (isAllowedSectionPath(config, document.docType, block.sections)) {
        if (rule.allowInAllowedSections) {
          continue;
        }
        if (severity === "off" || severity === "suggestion") {
          continue;
        }
      }

      if (severity === "off") {
        continue;
      }

      const [line, column] = locateMatch(block, match.index);
      findings.push({
        path: document.path,
        docType: document.docType,
        line,
        column,
        rule: rule.id,
        severity,
        message: rule.message,
        excerpt: block.text.trim(),
        section: block.section,
        sections: block.sections,
      });
    }
  }
  return findings;
}

export function* iterProseBlocks(text: string): Iterable<ProseBlock> {
  const sectionStack: Array<[number, string]> = [];
  let segments: Segment[] = [];
  let inFence = false;

  const flush = (): ProseBlock | undefined => {
    if (segments.length === 0) {
      return undefined;
    }
    const normalizedSegments: Segment[] = [];
    let offset = 0;
    for (const segment of segments) {
      normalizedSegments.push({ ...segment, start: offset });
      offset += segment.text.length + 1;
    }
    const block: ProseBlock = {
      text: normalizedSegments.map((segment) => segment.text).join(" "),
      segments: normalizedSegments,
      sections: sectionStack.map(([, title]) => title),
    };
    block.section = block.sections.at(-1);
    segments = [];
    return block;
  };

  const lines = text.split(/\r?\n/);
  for (let index = 0; index < lines.length; index += 1) {
    const lineNumber = index + 1;
    const line = lines[index];
    const stripped = line.trim();

    if (FENCE_RE.test(line)) {
      const block = flush();
      if (block) {
        yield block;
      }
      inFence = !inFence;
      continue;
    }

    if (inFence) {
      continue;
    }

    const heading = HEADING_RE.exec(line);
    if (heading) {
      const block = flush();
      if (block) {
        yield block;
      }
      const level = heading[1].length;
      const title = stripMarkdownTitle(heading[2]);
      while (sectionStack.length > 0 && (sectionStack.at(-1)?.[0] ?? 0) >= level) {
        sectionStack.pop();
      }
      sectionStack.push([level, title]);
      continue;
    }

    if (!stripped || looksLikeTableRow(stripped)) {
      const block = flush();
      if (block) {
        yield block;
      }
      continue;
    }

    const leading = line.length - line.trimStart().length;
    segments.push({ line: lineNumber, column: leading + 1, start: 0, text: stripped });
  }

  const block = flush();
  if (block) {
    yield block;
  }
}

function locateMatch(block: ProseBlock, offset: number): [number, number] {
  for (const segment of block.segments) {
    const end = segment.start + segment.text.length;
    if (segment.start <= offset && offset <= end) {
      return [segment.line, segment.column + Math.max(0, offset - segment.start)];
    }
  }
  const last = block.segments.at(-1);
  return last ? [last.line, last.column] : [1, 1];
}

function looksLikeTableRow(stripped: string): boolean {
  if (stripped.startsWith("|") && stripped.endsWith("|")) {
    return true;
  }
  return /^\|?\s*:?-{3,}:?\s*(\|\s*:?-{3,}:?\s*)+\|?$/.test(stripped);
}

function stripMarkdownTitle(title: string): string {
  return title
    .trim()
    .replace(/^#+|#+$/g, "")
    .trim();
}
